### PR TITLE
Ignore basepython for default factors (#477)

### PR DESCRIPTION
**Pull request by stephenfin**
tox provides a number of default factors - py27, py34, py35 etc. - that
are tied to particular interpreter versions. It is possible to override
these through individual sections or the global [testenv] section. For
example, consider the following 'tox.ini' file:

	[tox]
	skipsdist = True
	minversion = 2.0
	distribute = False
	envlist = py35,py27,pep8,py34-test

	[testenv]
	basepython = python3
	install_command = pip install {opts} {packages}
	commands =
	python --version

	[testenv:py27]
	basepython = python2.7

Running any target except for 'py27' will result in the same interpreter
being used. On Fedora 28 with the 'python3-tox' package:

	$ tox -qq -e py27
	Python 2.7.15
	$ tox -qq -e py35
	Python 3.6.5
	$ tox -qq -e py34-test
	Python 3.6.5

This is broken by design. Overriding these makes no sense and is a
source of common misconfigurations, as noted in #477. The only sane
thing to do here is ignore the request and use the correct interpreter
so this is what's done.

### CODE REVIEW

**Comment by gaborbernat** [comment]
@stephenfin thank you very much for the configuration. I would prefer to raise an error for invalid configuration though; instead of ignoring it. @obestwalter agree?

**Comment by stephenfin** [comment]
> @stephenfin thank you very much for the configuration. I would prefer to raise an error for invalid configuration though; instead of ignoring it. @obestwalter agree?

I considered that but decided not to because I wanted to allow configurations like the following:

```
[tox]
envlist = py27,py35,py36,docs,style,coverage

[testenv]
basepython = python3
commands = pytest ...

[testenv:docs]
commands = sphinx-build ...

[testenv:coverage]
commands = coverage ...

[testenv:style]
commands = flake8 ...
```

e.g. I would be able to ensure everything *except* the Python-versioned environments would use Python 3, to assist in migrations to Python 3. This has come up repeatedly in OpenStack, as seen [here](https://review.openstack.org/#/c/571207/1/tox.ini) and [here](https://review.openstack.org/#/q/topic:python3-first), and simply overriding things allows us to keep our `tox.ini` smaller. I would have added a DEBUG-level log, but it seems you're not using any logging infrastructure so I can't.

Your call.

**Comment by gaborbernat** [comment]
@stephenfin I see your point; I say let's follow the zen of Python: explicit better than implicit. So instead of blindly ignoring things let's have a way to opt-in into such decisions. 

```
basepython_env_name_missmatch = raise | ignore | override
```

Where raise aborts with a config error; ignore has the current strategy, and override can be what you're proposing here. Let's leave ``ignore`` as default for now (we can change it when we do a major release), and we should also expose this option as a command line argument to facilitate with testing/debugging. Does that seem reasonable to you?

**Comment by stephenfin** [comment]
> @stephenfin I see your point; I say let's follow the zen of Python: explicit better than implicit. So instead of blindly ignoring things let's have a way to opt-in into such decisions.
> 
>     basepython_env_name_missmatch = raise | ignore | override
> 
> Where raise aborts with a config error; ignore has the current strategy, and override can be what you're proposing here. Let's leave ignore as default for now (we can change it when we do a major release), and we should also expose this option as a command line argument to facilitate with testing/debugging. Does that seem reasonable to you?

That sounds mostly OK but I'd like to propose a slight tweak.

    ignore_basepython_conflict = True | False (default)

Where `False` preserves the current behaviour in this version but raises a warning and `True` gives me the behaviour I described above. Once tox 4.0 is released, we could bump this warning to an exception. What do you think?

**Comment by gaborbernat** [comment]
Ok. Make the changes and we can roll with it.

File: doc/conf.py [hunk 1]
```diff
@@ -47,7 +47,7 @@
 def setup(app):
     # from sphinx.ext.autodoc import cut_lines
     # app.connect('autodoc-process-docstring', cut_lines(4, what=['module']))
-    app.add_description_unit(
+    app.add_object_type(
```
Comment (gaborbernat) [inline doc/conf.py]: please explain this change :+1:

File: doc/config.rst [hunk 2]
```diff
@@ -81,12 +93,19 @@ Complete list of settings that you can put into ``testenv*`` sections:
...
+
+    .. versionchanged:: 3.1
+
+       Environments that use a :ref:`default factor <factors>` now ignore this
+       value, defaulting to the interpreter defined for that factor.
 
 .. confval:: commands=ARGVLIST
 
-    the commands to be called for testing. Each command is defined
+    The commands to be called for testing. Each command is defined
```
Comment (gaborbernat) [inline doc/config.rst]: please explain this change :+1:

File: doc/config.rst [hunk 3]
```diff
@@ -114,38 +133,36 @@ Complete list of settings that you can put into ``testenv*`` sections:
...
-
 .. confval:: ignore_errors=True|False(default)
 
     .. versionadded:: 2.0
 
-      If ``True``, a non-zero exit code from one command will be ignored and
-      further commands will be executed (which was the default behavior in tox <
-      2.0).  If ``False`` (the default), then a non-zero exit code from one command
-      will abort execution of commands for that environment.
+    If ``True``, a non-zero exit code from one command will be ignored and
```
Comment (gaborbernat) [inline doc/config.rst]: please explain this change :+1:

File: doc/config.rst [hunk 4]
```diff
@@ -205,51 +223,53 @@ Complete list of settings that you can put into ``testenv*`` sections:
...
-   .. versionadded:: 2.0
+    .. versionadded:: 2.0
 
-   A list of wildcard environment variable names which
-   shall be copied from the tox invocation environment to the test
-   environment when executing test commands.  If a specified environment
-   variable doesn't exist in the tox invocation environment it is ignored.
-   You can use ``*`` and ``?`` to match multiple environment variables with
-   one name.
+    A list of wildcard environment variable names which
```
Comment (gaborbernat) [inline doc/config.rst]: please explain this change :+1:

File: doc/config.rst [hunk 5]
```diff
@@ -205,51 +223,53 @@ Complete list of settings that you can put into ``testenv*`` sections:
...
+        in ``passenv`` a warning will be emitted.
 
 .. confval:: recreate=True|False(default)
 
     Always recreate virtual environment if this option is True.
 
 .. confval:: downloadcache=path
 
-    **IGNORED** -- Since pip-8 has caching by default this option is now ignored.  Please remove it from your configs as a future tox version might bark on it.
+    **IGNORED** -- Since pip-8 has caching by default this option is now
```
Comment (gaborbernat) [inline doc/config.rst]: please explain this change :+1:

File: src/tox/config.py [hunk 6]
```diff
@@ -505,11 +505,32 @@ def setenv(testenv_config, value):
...
+        for factor in testenv_config.factors:
+            if factor in tox.PYTHON.DEFAULT_FACTORS:
+                default = tox.PYTHON.DEFAULT_FACTORS[factor]
+
+                if value is None or testenv_config.config.ignore_basepython_conflict:
+                    return default
+
+                if str(value) != default:
+                    # TODO(stephenfin): Raise an exception here in tox 4.0
+                    print(
```
Comment (gaborbernat) [inline src/tox/config.py]: let's use the warning system instead, https://docs.python.org/2/library/warnings.html; as the user can turn off that if they want

File: tox.ini [hunk 7]
```diff
@@ -25,9 +25,7 @@ description = invoke sphinx-build to build the HTML docs and check that all link
 whitelist_externals = sphinx-build
 basepython = python3.6
 extras = docs
-changedir = {toxinidir}
```
Comment (gaborbernat) [inline tox.ini]: why did you remove this, I think that's why things fail now

File: src/tox/config.py (hunk 6 above)
Comment (stephenfin) [reply src/tox/config.py]: Sounds good. I wasn't able to find any other examples of warnings in use so I assumed you didn't use them.

**Comment by stephenfin** [comment]
@gaborbernat I have a bad habit of using GitHub like I use mailing lists and Gerrit (including force pushes) :smile: The answer to all your questions above is in the [commit messages](https://github.com/tox-dev/tox/pull/841/commits). Is this what you're looking for?

File: doc/conf.py (hunk 1 above)
Comment (stephenfin) [reply doc/conf.py]: See https://github.com/tox-dev/tox/pull/841/commits/116afb14d12fc16e2f7a98b685bd7c7bab2e125c

File: doc/config.rst (hunk 2 above)
Comment (stephenfin) [reply doc/config.rst]: See https://github.com/tox-dev/tox/pull/841/commits/f52eea73ba00094c46979851551c74fc3e7de9ec

**Comment by gaborbernat** [comment]
Yeah, migrate to the warnings api and we're good.

**Comment by stephenfin** [comment]
> Yeah, migrate to the warnings api and we're good.

This is done now.

**Comment by gaborbernat** [comment]
@stephenfin thank you very much!
